File: old/old_xml-and-json/src/main/java/org/tsdes/spring/xmlandjson/Converter.java
package org.tsdes.spring.xmlandjson;



public interface Converter<T> {

    String toXML(T obj);

    T fromXML(String xml);

    String toJSon(T obj);

    T fromJSon(String json);


    /*
        Transforming strings directly from XML to JSON (and vice-versa) would
        be more efficient. But piping together existing methods is very quick
        and easy, and perfectly valid solution when performance is not a major
        constraint.

        Recall that these default methods in this interface could be overwritten
        if necessary
     */


    default String fromJSonToXml(String json){

        T obj = fromJSon(json);
        String xml = toXML(obj);

        return xml;
    }

    default String fromXmlToJSon(String xml){

        T obj = fromXML(xml);
        String json = toJSon(obj);

        return json;
    }
}

File: advanced/data-format/src/main/kotlin/org/tsdes/advanced/dataformat/Converter.kt
package org.tsdes.advanced.dataformat


/**
 * Note: by default, in Kotlin a generic T would bind to "Any?", which
 * allows nullable types (eg, "String?").
 * If we want to prevent null, we need "T: Any"
 */
interface Converter<T: Any> {

    fun toXml(obj: T): String

    fun fromXml(xml: String): T

    fun toJson(obj: T): String

    fun fromJson(json: String): T


    /*
        Transforming strings directly from XML to JSON (and vice-versa) would
        be more efficient. But piping together existing methods is very quick
        and easy, and perfectly valid solution when performance is not a major
        constraint.

        Recall that these default methods in this interface could be overwritten
        if necessary
     */


    fun fromJsonToXml(json: String): String {

        val obj = fromJson(json)

        return toXml(obj)
    }

    fun fromXmlToJson(xml: String): String {

        val obj = fromXml(xml)

        return toJson(obj)
    }
}